### DISCUSSION TITLE
avoiding to make custom hooks while using useShallow

### DISCUSSION THREAD

**Discussion by ImamJanjua** [pmndrs/zustand#2547] [General]
Hi, am trying to create an api, like in this file I have a main function and a helper function and both functions need some data from the store to execute. And than I wanna put the main func in the useMutation hook from react query. 
But the problem ofcourse is that by calling the state from the store in this way, react will complain. Solution kinda would be custom hook but that would make the structure more complicated. 
I could select the the states without the useShallow hook but wouldn't i than loose the render optimization?

```typescript
const [id] = useDoctorStore(useShallow((state) => [state.id]));


async function helperFunc() {
  //...
}

export async function mainFunc() {
  //...
}

```

**Comment by dai-shi** [comment]
If you don't prefer making a custom hook, how about using `createWithEqualityFn` with `shallow`?

```js
import { createWithEqualityFn } from 'zustand/traditional';
import { shallow } from 'zustand/shallow';

export const useDoctorStore = createWithEqualitfyFn((set) => ({
  // ...
}), shallow);
```

**Reply by ImamJanjua** [reply]
thanks! but why is this no where in the docs?

**Reply by dbritto-dev** [reply]
@ImamJanjua we are working on revamping docs

**Reply by ImamJanjua** [reply]
@dbritto-dev  Thanks for the response! But the thing as i mentioned that i wanna create a helper function and pass it as `mutationFn` to `useMutation`
 from react-query.

This function writes to backend and should also update the depended store

Initially I wanted to call the needed states and actions  at top of the file so that the functions need them can use them, but that ofcourse would not work since `useFooStore` istself is a hook. How would u approach that ?

**Reply by dbritto-dev** [reply]
@ImamJanjua in that case you want to use the store itself the set state function

**Reply by ImamJanjua** [reply]
@dbritto-dev So put a action for that in the store? Didn't got that.

For clarification, wanted to do something like that: 
```typescript
import { doc, setDoc } from "firebase/firestore";
import { FIRESTORE } from "@/firebase/config";
import { useMutation } from "@tanstack/react-query";

export async function changeValue(value: string) {
  // writes to backend
  const docRef = doc(FIRESTORE, "test", "test");
  await setDoc(docRef, { value });

  // update store
  //........
}

export const useChangeValue = () =>
  useMutation({
    mutationFn: changeValue,
    onError: (error) => {
      console.error("Error changing value:", error);
    },
  });
```

**Reply by dbritto-dev** [reply]
@ImamJanjua yeah, that and then update the store with the result of your request

**Reply by ImamJanjua** [reply]
@dbritto-dev  that what i meant, will complain because using a hook in normal function

```typescript
import { doc, setDoc } from "firebase/firestore";
import { FIRESTORE } from "@/firebase/config";
import { useMutation } from "@tanstack/react-query";

import { useTestStore } from "@/stores/testStore";

export async function changeValue(value: string) {
  const setValue = useTestStore((state) => state.setValue);

  // writes to backend
  const docRef = doc(FIRESTORE, "test", "test");
  await setDoc(docRef, { value });

  // upadtes store
  setValue(value);
}

export const useChangeValue = () =>
  useMutation({
    mutationFn: changeValue,
    onError: (error) => {
      console.error("Error changing value:", error);
    },
  });
```

**Reply by dbritto-dev** [reply]
@ImamJanjua you should be able to use the store API utilities like `getState`. I mean this: 

```ts
const setValue = useTestStore.getState().setValue
```

**Reply by ImamJanjua** [reply]
@dbritto-dev btw, what the benifit of this custom code in docs for selectors, If I can call states and actions with the hooks or with the store API utilities?

**Reply by dbritto-dev** [reply]
@ImamJanjua you are mixing two things selectors and store API utilities like `getState`. The store API utilities are a way to interact with the store. In the other hand selector functions and equality functions are ways to optimize re renders are strictly associated with React

**Reply by ImamJanjua** [reply]
@dbritto-dev  i got that, but i mean whats the benifit of using const bear = useBearStore.use.bear() instead of const nuts = useBearStore((state) => state.nuts) ?

**Reply by dbritto-dev** [reply]
> @dbritto-dev  i got that, but i mean whats the benifit of using const bear = useBearStore.use.bear() instead of const nuts = useBearStore((state) => state.nuts) ?

@ImamJanjua it's the same thing, you are making them less verbose but that's all

**Reply by ImamJanjua** [reply]
@dbritto-dev  Thanks, but why is it than in the docs?

**Reply by dbritto-dev** [reply]
@ImamJanjua what you mean? they are just different ways to do the same

**Reply by ImamJanjua** [reply]
@dbritto-dev, do u know why the function can only gets the newest states when i refresh?


https://github.com/pmndrs/zustand/assets/137432044/c82b0248-4f96-4b22-9bee-0935bc7da53b

**Reply by dbritto-dev** [reply]
@ImamJanjua you should use `useTestStore.getState()` inside of your `changeValue()` function, having that logic outside of that function is the reason of what you are facing that issue
